File: Cap 10/Topico 10.6/Topico 10.6.2/calcfat2.pas
{*
 * Livro .....: Algoritmos
 *              Logica para Desenvolvimento de Programacao de
 *              Computadores
 * Autores ...: Jose Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Erica
 *
 * Capitulo ..: 10 - Topico 10.6.2
 * 
 * Programa ..: PASSAGEM DE PARAMETRO POR REFERENCIA
 * 
 *}

program CALC_FAT_V2;

procedure FATORIAL(N: byte; var FAT: qword);
var
  I: byte;
begin
  FAT := 1;
  for I := 1 to N do
    FAT := FAT * I;
end;

var
  LIMITE: byte;
  RESP: qword;

begin

  RESP := 1;
  writeln('CALCULO DE FATORIAL');
  writeln('PROCEDIMENTO COM PASSAGEM DE PARAMETRO POR REFERENCIA');
  writeln;
  
  write('Qual fatorial (1-20): ');
  readln(LIMITE);

  FATORIAL(LIMITE, RESP);
  writeln('Fatorial = ', RESP);
  
end.

File: Cap 10/10.6 - Passagens de Parametro/10.6.2 - Passagem de Parametro por Referencia/Simulado/calcfat2.rb
=begin
 * Livro .....: Algoritmos
 *              Logica para Desenvolvimento de Programacao de
 *              Computadores
 * Autores ...: Jose Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Erica
 *
 * Capitulo ..: 10 - Topico 10.6.2
 * 
 * Programa ..: PASSAGEM DE PARAMETRO POR REFERENCIA (SIMULADA)
=end

def fatorial(n, fat)
  for i in 1..n
    fat[0] = fat[0] * i
  end
end

limite = 0
resp = [1]

puts 'CALCULO DE FATORIAL'
puts 'PROCEDIMENTO COM PASSAGEM DE PARAMETRO POR REFERENCIA'
puts

print 'Qual fatorial (1-N): '
limite = gets.chomp.to_i

fatorial(limite, resp)
puts "Fatorial = #{resp[0]}"
